Source/Core/Common/StdOptional.h

```diff
@@ -0,0 +1,913 @@
+  }
+  constexpr const T* dataptr() const
+  {
+    return detail_::static_addressof(OptionalBase<T>::storage_.value_);
+  }
+
+  constexpr const T& contained_val() const& { return OptionalBase<T>::storage_.value_; }
+  constexpr T&& contained_val() && { return std::move(OptionalBase<T>::storage_.value_); }
+  constexpr T& contained_val() & { return OptionalBase<T>::storage_.value_; }
+  void clear() noexcept
+  {
+    if (initialized())
+      dataptr()->T::~T();
+    OptionalBase<T>::init_ = false;
+  }
+
+  template <class... Args>
+  void initialize(Args&&... args) noexcept(noexcept(T(std::forward<Args>(args)...)))
+  {
+    assert(!OptionalBase<T>::init_);
+    ::new (static_cast<void*>(dataptr())) T(std::forward<Args>(args)...);
+    OptionalBase<T>::init_ = true;
+  }
+
+  template <class U, class... Args>
+  void initialize(std::initializer_list<U> il,
+                  Args&&... args) noexcept(noexcept(T(il, std::forward<Args>(args)...)))
+  {
+    assert(!OptionalBase<T>::init_);
+    ::new (static_cast<void*>(dataptr())) T(il, std::forward<Args>(args)...);
+    OptionalBase<T>::init_ = true;
+  }
+
+public:
+  typedef T value_type;
+
+  // 20.5.5.1, constructors
+  constexpr optional() noexcept : OptionalBase<T>(){};
+  constexpr optional(nullopt_t) noexcept : OptionalBase<T>(){};
+
+  optional(const optional& rhs) : OptionalBase<T>()
+  {
+    if (rhs.initialized())
+    {
+      ::new (static_cast<void*>(dataptr())) T(*rhs);
+      OptionalBase<T>::init_ = true;
+    }
+  }
+
+  optional(optional&& rhs) noexcept(is_nothrow_move_constructible<T>::value) : OptionalBase<T>()
+  {
+    if (rhs.initialized())
+    {
+      ::new (static_cast<void*>(dataptr())) T(std::move(*rhs));
+      OptionalBase<T>::init_ = true;
+    }
+  }
+
+  constexpr optional(const T& v) : OptionalBase<T>(v) {}
+  constexpr optional(T&& v) : OptionalBase<T>(constexpr_move(v)) {}
+  template <class... Args>
+  explicit constexpr optional(in_place_t, Args&&... args)
+      : OptionalBase<T>(in_place_t{}, constexpr_forward<Args>(args)...)
+  {
+  }
+
+  template <class U, class... Args,
+            TR2_OPTIONAL_REQUIRES(is_constructible<T, std::initializer_list<U>>)>
+  constexpr explicit optional(in_place_t, std::initializer_list<U> il, Args&&... args)
+      : OptionalBase<T>(in_place_t{}, il, constexpr_forward<Args>(args)...)
+  {
+  }
+
+  // 20.5.4.2, Destructor
+  ~optional() = default;
+
+  // 20.5.4.3, assignment
+  optional& operator=(nullopt_t) noexcept
+  {
+    clear();
+    return *this;
+  }
+
+  optional& operator=(const optional& rhs)
+  {
+    if (initialized() && !rhs.initialized())
+      clear();
+    else if (!initialized() && rhs.initialized())
+      initialize(*rhs);
+    else if (initialized() && rhs.initialized())
+      contained_val() = *rhs;
+    return *this;
+  }
+
+  optional& operator=(optional&& rhs) noexcept(
+      is_nothrow_move_assignable<T>::value&& is_nothrow_move_constructible<T>::value)
```

Comment:
This is what clang-format wanted to format it to (looks like a bug). I have replaced these ::value with (), which incidentally fixes the missing space issue.